File: mobile/lib/features/customers/domain/entities/invoice.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'invoice.freezed.dart';
part 'invoice.g.dart';

enum InvoiceStatus {
  @JsonValue('unpaid')
  unpaid,
  @JsonValue('partial')
  partial,
  @JsonValue('paid')
  paid,
  @JsonValue('cancelled')
  cancelled,
}

@freezed
abstract class Invoice with _$Invoice {
  const factory Invoice({
    required String id,
    @JsonKey(name: 'customer_id') required String customerId,
    @JsonKey(name: 'deal_id') String? dealId,
    @JsonKey(name: 'invoice_no') required String invoiceNo,
    required double amount,
    @JsonKey(name: 'paid_amount') required double paidAmount,
    required InvoiceStatus status,
    @JsonKey(name: 'due_at') DateTime? dueAt,
    @JsonKey(name: 'signature_path') String? signaturePath,
    @JsonKey(name: 'created_at') DateTime? createdAt,
    @JsonKey(name: 'updated_at') DateTime? updatedAt,
    @Default([]) List<InvoiceItem> items,
  }) = _Invoice;

  factory Invoice.fromJson(Map<String, dynamic> json) => _$InvoiceFromJson(json);
}

@freezed
abstract class InvoiceItem with _$InvoiceItem {
  const factory InvoiceItem({
    required String id,
    @JsonKey(name: 'invoice_id') required String invoiceId,
    @JsonKey(name: 'product_id') required String productId,
    required String name,
    required double quantity,
    required String unit,
    @JsonKey(name: 'unit_price') required double unitPrice,
    required double subtotal,
  }) = _InvoiceItem;

  factory InvoiceItem.fromJson(Map<String, dynamic> json) => _$InvoiceItemFromJson(json);
}

File: backend/internal/domain/models/invoice.go
package models

import (
	"crm_wowin_backend/pkg/utils"
	"github.com/google/uuid"
)

type InvoiceStatus string

const (
	InvoiceStatusUnpaid    InvoiceStatus = "unpaid"
	InvoiceStatusPartial   InvoiceStatus = "partial"
	InvoiceStatusPaid      InvoiceStatus = "paid"
	InvoiceStatusCancelled InvoiceStatus = "cancelled"
)

type Invoice struct {
	ID            uuid.UUID       `json:"id"`
	CustomerID    uuid.UUID       `json:"customer_id"`
	DealID        *uuid.UUID      `json:"deal_id,omitempty"`
	InvoiceNo     string          `json:"invoice_no"`
	Amount        float64         `json:"amount"`
	PaidAmount    float64         `json:"paid_amount"`
	Status        InvoiceStatus   `json:"status"`
	SignaturePath string          `json:"signature_path,omitempty"`
	DueAt         *utils.FlexTime `json:"due_at,omitempty"`
	CreatedAt     utils.FlexTime  `json:"created_at"`
	UpdatedAt     utils.FlexTime  `json:"updated_at"`
	
	// Detail items (optional)
	Items []*InvoiceItem `json:"items,omitempty"`
}

type InvoiceItem struct {
	ID        uuid.UUID `json:"id"`
	InvoiceID uuid.UUID `json:"invoice_id"`
	ProductID uuid.UUID `json:"product_id"`
	Name      string    `json:"name"`
	Quantity  float64   `json:"quantity"`
	Unit      string    `json:"unit"`
	UnitPrice float64   `json:"unit_price"`
	Subtotal  float64   `json:"subtotal"`
    CreatedAt utils.FlexTime `json:"created_at"`
}
